# Name: Ejegu Smith
# Class: DTSC 680
# Assignment: 6

In [1]:
import pandas as pd 
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

In [2]:
col_names = ['class','cap-shape', 'cap-surface', 'cap-color', 'bruises?', 'odor', 'gill-attachment', 'gill-spacing',
            'gill-size', 'gill-color', 'stalk-shape', 'stalk-root', 'stalk-surface-above-ring', 
             'stalk-surface-below-ring', 'stalk-color-above-ring',  'stalk-color-below-ring',
            'veil-type', 'veil-color', 'ring-number', 'ring-type', 'spore-print-color',
            'population', 'habitat']

df = pd.read_csv('agaricus-lepiota.data', names=col_names)
df.shape

(8124, 23)

Next, you should train your KNN model on those instances that don’t have missing values, then have the model make predictions for those instances that have missing values. This is how you will have the KNN model impute missing data

In [3]:
#
df_no_missing = df[df['stalk-root'] != '?']
df_no_missing.shape

(5644, 23)

In [4]:
missing_values_df = df[df['stalk-root'] == '?']
missing_values_df.shape

(2480, 23)

In [5]:
lab_enc_missing = LabelEncoder()
response_missing = lab_enc_missing.fit_transform(missing_values_df['stalk-root'])

one_hot_missing = OneHotEncoder(sparse = False)
features_missing = one_hot_missing.fit_transform(missing_values_df.drop(['stalk-root'], axis=1))

In [6]:
features_missing.shape

(2480, 83)

In [7]:
#trying to predict missing values which means stalk-root is going to be response data, and everythiong else is features 
#one hot encode and label encode
#training data is all the rows that do not have ? mark in stalk root
#testing data are the remainder rows
# array of (mask index) true and false
#identifty what rows have question mark in stalk root and vise versa**


**I OHE and label encoded the entire data set, then proceeded to train/test/split the '?' excluded DF into the knn model I trained. Then split the features/response '?' included DF and predicted using the features DF and it worked. 

Attempting to fit/fit_transform the two separate sets of data (exclude/include ?) was causing the dimensionality issue.

In [8]:
response_no_missing = df_no_missing[['stalk-root']]
features_no_missing = df_no_missing.drop(['stalk-root'], axis=1)

In [9]:
lab_enc_missing = LabelEncoder()
response_no_missing = lab_enc_missing.fit_transform(response_no_missing)

one_hot_no_missing = OneHotEncoder(sparse = False)
features_no_missing = one_hot_no_missing.fit_transform(features_no_missing)

/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


In [10]:
features_no_missing.shape

(5644, 96)

In [65]:
np.unique(response_no_missing)

array([0, 1, 2, 3])

In [11]:
#X_train, X_test, y_train, y_test = train_test_split(features, response, test_size=.20, random_state=42)

knn_clf1 = KNeighborsClassifier(n_neighbors=5)

knn_clf1.fit(features_no_missing, response_no_missing)

#predicting missing values for stalk-root
#stalk_root_preds = knn_clf1.predict(X_test)

#stalk_root_preds

KNeighborsClassifier()

In [12]:
#predicting missing values for stalk-root
stalk_root_preds = knn_clf1.predict(features_missing)

stalk_root_preds

ValueError: query data dimension must match training data dimension

In [13]:
stalk_root_preds.shape

NameError: name 'stalk_root_preds' is not defined

In [14]:
stalk_root_preds = lab_enc.inverse_transform(stalk_root_preds)
stalk_root_preds

NameError: name 'lab_enc' is not defined

In [15]:
lab_enc.classes_

NameError: name 'lab_enc' is not defined

In [16]:
##shape does not match missing values count for stalk root?
stalk_root_preds.shape

NameError: name 'stalk_root_preds' is not defined

In [17]:
missing_values = df[df['stalk-root'] == '?']
missing_values = missing_values[['stalk-root']]
missing_values

stalk-root
3984          ?
4023          ?
4076          ?
4100          ?
4104          ?
...         ...
8119          ?
8120          ?
8121          ?
8122          ?
8123          ?

[2480 rows x 1 columns]

In [18]:
import numpy as np
missing_values.replace('?', np.NaN, inplace=True)
missing_values

stalk-root
3984         NaN
4023         NaN
4076         NaN
4100         NaN
4104         NaN
...          ...
8119         NaN
8120         NaN
8121         NaN
8122         NaN
8123         NaN

[2480 rows x 1 columns]

## Now making models to predict for class (edible or poisonous)...

In [19]:
#after i imput data then use features and response
#features = df_no_missing.drop(['class'], axis=1)
#response = df_no_missing[['class']]

In [20]:
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

In [21]:
response = df_no_missing[['stalk-root']]
features = df_no_missing.drop(['stalk-root'], axis=1)

In [22]:
#label encode response data
#1 = posionous
#0 = edible


lab_enc = LabelEncoder()
response = lab_enc.fit_transform(response)
response

/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


array([2, 1, 1, ..., 0, 1, 1])

In [23]:
#checking
lab_enc.classes_

array(['b', 'c', 'e', 'r'], dtype=object)

In [24]:
#checking
response.shape

(5644,)

In [25]:
#now one hot encode catefgorical features df which are the categorical variables
one_hot = OneHotEncoder(sparse = False)
features = one_hot.fit_transform(features)
features

array([[0., 1., 0., ..., 0., 0., 1.],
       [1., 0., 0., ..., 0., 0., 0.],
       [1., 0., 1., ..., 1., 0., 0.],
       ...,
       [1., 0., 0., ..., 0., 1., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.]])

In [26]:
#now splitting test and training data
X_train, X_test, y_train, y_test = train_test_split(features, response, test_size=.20, random_state=42)

In [27]:
# creating and training the KNN model, then preforming predictions
knn_clf = KNeighborsClassifier(n_neighbors=5, random_state=42)

knn_clf.fit(X_train, y_train)

preds = knn_clf.predict(X_test)

In [28]:
from sklearn.metrics import accuracy_score

#KNN model accuracy score
accuracy_score(preds, y_test)

1.0

## Graded Concept Question #1

 Would it still be
possible to train the KNN model if you one-hot encoded the response data instead?
Why or why not?

Yes, it would still be possible as machine learning models need numerical inputs. In this case, the target data is either e for edible or p for poosionus. Therefore the target data must be converted to 1s and 0s. Using one-hot encoder on the target column would still result in the data being transformed from categorical to numerical, and thus the KNN model would be able to accept the target data.Furthermore, the target column is nominal instead of oridinal (no logical ordering), therefore one hot encoder can be used on the column. 

 3.Train a RandomForestClassifier as well as a LogisticRegression model to predict
whether a mushroom is edible or poisonous given this data set of nominally-valued
characteristics. Train the model on the feature data supplied to you after you’ve one-hot
encoded it. You should label encode the response data.

In [29]:
from sklearn.ensemble import RandomForestClassifier

rnd_clf = RandomForestClassifier(random_state=42)

rnd_clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [30]:
y_pred_rnd_clf = rnd_clf.predict(X_test)
y_pred_rnd_clf

array([2, 0, 0, ..., 1, 0, 0])

In [31]:
#random forest classifier accuracy score
accuracy_score(y_pred_rnd_clf, y_test)

1.0

In [49]:
from sklearn.linear_model import LogisticRegression

log_clf = LogisticRegression(random_state=42)

log_clf.fit(X_train, y_train)

/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

In [50]:
y_preds_log_clf = log_clf.predict(X_test)
y_preds_log_clf

array([2, 0, 0, ..., 1, 0, 0])

In [51]:
#Logistic Regression classifier accuracy score
accuracy_score(y_preds_log_clf, y_test)

1.0

## Graded Concept Question #2

Would it still be possible to train these two models if you one-hot encoded the response data instead,
being careful to specify that the drop parameter of the OneHotEncoder class is set to
‘first’? Why or why not?

No, it isn't possible. I recieved the following ValueError: "sparse multilabel-indicator for y is not supported." This is due to the curse of dimensionality problem. Additionally, the models do not support multiple columns for y, which is the result of applying the one hot encoder (ie one coulmn for edible and one column for posionous each with rows of 1's and 0's ). 

4. When you train both the RandomForestClassifier and LogisticRegression models,
use the magic command %%time to time how long it takes to complete training. See the
Python Data Science Handbook for more information about this magic command.

In [57]:
print('Logistic Regression time:')
%time log_clf.fit(X_train, y_train)

Logistic Regression time:
CPU times: user 441 ms, sys: 30.5 ms, total: 472 ms
Wall time: 282 ms


/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

In [56]:
print('Random Forest Classifier time:')
%time rnd_clf.fit(X_train, y_train)

Random Forest Classifier time:
CPU times: user 279 ms, sys: 9.77 ms, total: 288 ms
Wall time: 344 ms


RandomForestClassifier(random_state=42)

5. Compute the accuracy, precision, and recall scores for a test set. Briefly discuss the
performance of your models in terms of these values.

In [62]:
#random forest classifier scores
from sklearn.metrics import recall_score, precision_score

acc_score = accuracy_score(y_pred_rnd_clf, y_test)
rec_score = recall_score(y_test, y_pred_rnd_clf, average="micro")
prec_score = precision_score(y_test, y_pred_rnd_clf, average="micro")

print("Accuracy score:", acc_score)
print('Recall score:', rec_score)
print('Precision score:', prec_score)

Accuracy score: 1.0
Recall score: 1.0
Precision score: 1.0


In [63]:
#logistic regression scores

acc_score = accuracy_score(y_preds_log_clf, y_test)
rec_score = recall_score(y_test, y_preds_log_clf, average="micro")
prec_score = precision_score(y_test, y_preds_log_clf, average="micro")

print("Accuracy score:", acc_score)
print('Recall score:', rec_score)
print('Precision score:', prec_score)

Accuracy score: 1.0
Recall score: 1.0
Precision score: 1.0


In [ ]:
# All values are 100%, which means the models are severly over-fitting. This could be due to 

6. Perform dimensionality reduction using PCA and keep 95% of the variance. By what
percentage were you able to reduce the number of dimensions of the training set? How
many features (i.e. dimensions) are you left with after reducing dimensionality?

In [38]:
from sklearn.decomposition import PCA
import numpy as np

pca = PCA(n_components=5)
X_reduced = pca.fit_transform(X_train)

In [39]:
#25% of the datasets variance lies on the 1st axis, and 12% on the 2nd axis, and so on
pca.explained_variance_ratio_

array([0.24937043, 0.11855212, 0.09218185, 0.0454819 , 0.04458208])

In [40]:
pca2 = PCA()
pca2.fit(X_train)
cumsum = np.cumsum(pca2.explained_variance_ratio_)
optimal_dimensions = np.argmax(cumsum >= 0.95) + 1 #plus 1 becuase index starts at 0 
optimal_dimensions

33

In [41]:
pca2 = PCA(n_components=33)
X_reduced = pca2.fit_transform(X_train)

In [45]:
pca2.explained_variance_ratio_

array([0.24937043, 0.11855212, 0.09218185, 0.0454819 , 0.04458208,
       0.03642783, 0.03190587, 0.03046501, 0.02621419, 0.02248525,
       0.02013047, 0.01845981, 0.01728815, 0.01543633, 0.01457357,
       0.01374218, 0.0130995 , 0.01250255, 0.01203141, 0.01125247,
       0.01066765, 0.01026912, 0.01000825, 0.00989723, 0.00958588,
       0.00847868, 0.00797633, 0.00746149, 0.00719247, 0.00670819,
       0.0065527 , 0.00622625, 0.00573491])

we are left with 33 dimensions, this includes the expanded dimension data set due to one hot encoding columns

7. Train two new models, a RandomForestClassifier and a LogisticRegression model,
on this reduced dataset to predict whether a mushroom is edible or not. Again, time the
training of these two models on the reduced dataset. You will again use the default
hyperparameter values for both models, with a random state parameter of 42 for the
RandomForestClassifier.

In [54]:
rnd_clf_reduced = RandomForestClassifier(random_state=42)

%time rnd_clf_reduced.fit(X_reduced, y_train)

CPU times: user 1.16 s, sys: 10.7 ms, total: 1.17 s
Wall time: 1.19 s


RandomForestClassifier(random_state=42)

In [60]:
log_clf_reduced = LogisticRegression(random_state=42)

%time log_clf_reduced.fit(X_reduced, y_train)

CPU times: user 138 ms, sys: 30.5 ms, total: 168 ms
Wall time: 125 ms


LogisticRegression(random_state=42)

8. Compute the accuracy, precision, and recall scores for the model trained on the
reduced data set using the same test set as before. What conclusions can you make
about these models and the PCA process? How do the models compare? Discuss the
trends observed for model training time and performance for the full and reduced data
sets. Tabulate the following information in your final analysis:

                                               Random Forest:

| Item | Full Data | PCA Reduced |
|------|-----------|-------------|
|**Accuracy**|1.0|-------------|
|**Precision**|1.0|-------------|
|**Recall**|1.0|-------------|
|**Time**|288 ms|-------------|


                                               Logistic Regression:

| Item | Full Data | PCA Reduced |
|------|-----------|-------------|
|**Accuracy**|1.0|-------------|
|**Precision**|1.0|-------------|
|**Recall**|1.0|-------------|
|**Time**|472 ms|-------------|